skrifa/src/outline/autohint/shape.rs

```diff
@@ -592,6 +589,29 @@ impl<'a> GsubHandler<'a> {
         }
     }
 
+    /// Called when we begin processing a new lookup.
+    ///
+    /// Returns `false` if the lookup index is already being processed or if
+    /// the lookup would exceed our max depth.
+    fn enter_lookup(&mut self, lookup_index: u16) -> bool {
+        // Note: we just use a linear search here under the assumption that
+        // most fonts have shallow contextual lookup chains
+        if self.lookup_depth != 0
+            && (self.lookup_depth == MAX_NESTING_DEPTH
+                || self.lookup_stack[..self.lookup_depth].contains(&lookup_index))
+        {
+            return false;
+        }
+        self.lookup_stack[self.lookup_depth] = lookup_index;
+        self.lookup_depth += 1;
+        true
+    }
+
+    /// Called when we finish processing a single lookup.
+    fn exit_lookup(&mut self) {
+        self.lookup_depth = self.lookup_depth.saturating_sub(1);
+    }
```

Comment:
I don't love that this makes it relatively easy to add code that enters without calling exit. This would be somewhat harder if it was directly in the process lookup fn? - depicted below. Or by a guard that reduces depth on drop but I suppose that path takes you towards multiple mut refs.

```rust
    fn process_lookup(&mut self, lookup_index: u16) {
        // Guard: don't cycle and don't go exceed depth limit
        // Note: we use a linear search here under the assumption that
        // most fonts have shallow contextual lookup chains
        if self.lookup_depth != 0
            && (self.lookup_depth == MAX_NESTING_DEPTH
                || self.lookup_stack[..self.lookup_depth].contains(&lookup_index))
        {
            return;
        }
        self.lookup_stack[self.lookup_depth] = lookup_index;
        self.lookup_depth += 1;

        // Actually process the lookup
        self.process_lookup_inner(lookup_index);

        // Out we go again
        self.lookup_depth -= 1;
    }
```

Reply:
I agree with the sentiment but I kept the stack ops separate to support isolated unit testing of that functionality. If we embed them then we need to build a lookup list with the desired structure which seems less clear and more fragile to me.

Reply:
I would have thought we need to build lookups with good/bad structure (one that cycles, one that has max depth, one that has max depth + 1, etc) to unit test regardless so... that's kind of fine?